org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs

```diff
@@ -0,0 +1,718 @@
+                        
+                        WriteToLog(string.Format("    After Request (Label): {0}", watch.ElapsedMilliseconds));
+                    }
+                }
+                WriteToLog(string.Format("    End of Iteration: {0}", watch.ElapsedMilliseconds));
+                watch.Stop();
+            }
+
+            // Add contacts for people unknown to HubSpot
+            HashSet<int> hsPersonIds = new HashSet<int>();
+            HashSet<string> hsPersonEmails = new HashSet<string>();
+            foreach ( var contact in contacts )
+            {
+                hsPersonIds.Add( contact.properties.rock_person_id.AsInteger() );
+                hsPersonEmails.Add( contact.properties.email.ToStringSafe().ToLower() );
+            }
+            Debug.WriteLine( string.Format( "hsPersonIds is this big: {0}", hsPersonIds.Count ) );
+
+            int[] personIds = GetAllPersonIds(_context); // All Person Ids 18 years of age or older and is ACTIVE
+            foreach ( int id in personIds )
+            {
+                if ( !hsPersonIds.Contains( id ) )
+                {
+                    Debug.WriteLine(string.Format("hsPersonIds does not contain: {0}", id));
+                    Person person = personService.Get( id );
+                    if ( person != null )
+                    {
+                        if ( hsPersonEmails.Contains( person.Email.ToStringSafe().ToLower() ) ) { continue; } // If this person's email is already in HubSpot don't include the email with this newly added person.
+                        if ( hsPersonIds.Contains( person.Id ) ) { continue; } // redundant
+
+                        List<HubspotPropertyUpdate> properties = new List<HubspotPropertyUpdate>();
+                        properties.Add( new HubspotPropertyUpdate() { property = "firstname", value = person.NickName } );
+                        properties.Add( new HubspotPropertyUpdate() { property = "lastname", value = person.LastName } );
+                        properties.Add( new HubspotPropertyUpdate() { property = "rock_person_id", value = person.Id.ToString() } );
+                        PhoneNumber mobile = person.PhoneNumbers.FirstOrDefault( n => n.NumberTypeValueId == 12 );
+                        if ( mobile != null && !mobile.IsUnlisted && mobile.IsMessagingEnabled )
+                        {
+                            properties.Add( new HubspotPropertyUpdate() { property = "phone", value = mobile.NumberFormatted } );
+                        }
+
+                        string email = person.Email.ToLower();
+                        string[] emailSep = email.Split( '.' );
+                        string emailtld = "." + emailSep[emailSep.Length - 1];
+                        if ( person.CanReceiveEmail( true ) && hsPersonEmails.Contains( person.Email.ToStringSafe().ToLower() ) == false && email.Contains( "@" ) && tlds.Contains( emailtld ) )
```

Comment:
Why are we validating emails like this? This way of doing things leaves a few gaps. C# has a built-in way to validate emails that would be simpler and more consistent: [link](https://stackoverflow.com/questions/1365407/c-sharp-code-to-validate-email-address).